lib/ocpp/v201/smart_charging.cpp

```diff
@@ -120,6 +120,81 @@ std::ostream& operator<<(std::ostream& os, const ProfileValidationResultEnum val
     return os;
 }
 
+/// \brief Checks if the given filter \p criteria and \p evse_id matches the given \p profile
+bool profile_matches_clear_criteria(const ChargingProfile& profile, const std::optional<int32_t> profile_id,
+                                    const std::optional<ClearChargingProfile>& criteria, const int32_t evse_id) {
+    if (profile.chargingProfilePurpose == ChargingProfilePurposeEnum::ChargingStationExternalConstraints) {
+        // K10.FR.04; profiles with ChargingStationExternalConstraints shall not be removed
+        return false;
+    }
+
+    // // K10.FR.03, K10.FR.09
+    if (profile_id.has_value() and profile_id.value() == profile.id) {
+        return true;
+    }
+
+    if (profile_id.has_value()) {
+        // profile_id has value but doesnt match
+        return false;
+    }
+
+    if (!criteria.has_value()) {
+        // criteria has no value, so clear all
+        return true;
+    }
+
+    const auto _criteria = criteria.value();
+
+    if (_criteria.stackLevel.has_value() and _criteria.stackLevel.value() != profile.stackLevel) {
+        return false;
+    }
+    if (_criteria.evseId.has_value() and _criteria.evseId.value() != evse_id) {
+        return false;
+    }
+    if (_criteria.chargingProfilePurpose.has_value() and
+        _criteria.chargingProfilePurpose.value() != profile.chargingProfilePurpose) {
+        return false;
+    }
+    return true;
+}
+
+/// \brief Checks if the given filter \p criteria and \p evse_id matches the given \p profile
+bool profile_matches_get_criteria(const ChargingProfile& profile, const ChargingProfileCriterion& criteria,
+                                  const std::optional<int32_t> requested_evse_id, const int32_t profile_evse_id) {
+
+    if (criteria.chargingProfileId.has_value() and
+        std::find(criteria.chargingProfileId.value().begin(), criteria.chargingProfileId.value().end(), profile.id) !=
+            criteria.chargingProfileId.value().end()) {
+        return true;
+    }
+
+    if (criteria.chargingProfileId.has_value()) {
+        // no profile matches the id, no need to do check match of the criteria
+        return false;
+    }
+
+    if (criteria.chargingLimitSource.has_value() and
```

Comment:
Can refactor this into a function:  charging_limit_source_exists_and_is_not_CSO() 

or !charging_limit_source_exists_and_is_CSO()


And do similar refactorings for the rest of the if statements, so that it reads more like:

If (x || y || z || ...) {
   return false;
}

Reply:
I added a templated function `value_exists_and_is_different` to be able to reuse this in `profile_matches_clear_criteria` and `profile_matches_get_criteria`